NDTensors/src/dense/dense.jl

```diff
@@ -3,70 +3,101 @@
 #
 using LinearAlgebra: BlasFloat
 
-struct Dense{ElT,VecT<:AbstractArray} <: TensorStorage{ElT}
-  data::VecT
-  function Dense{ElT,VecT}(data::AbstractArray) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
+struct Dense{ElT,DataT<:AbstractArray} <: TensorStorage{ElT}
+  data::DataT
+  function Dense{ElT,DataT}(data::AbstractArray) where {ElT,DataT<:AbstractArray{ElT}}
+    return new{ElT,DataT}(data)
   end
 
-  # Special case of Vector
-  function Dense{ElT,VecT}(data::Vector) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
-  end
+  # # Special case of Vector
+  # function Dense{ElT,DataT}(data::Vector) where {ElT,DataT<:AbstractArray{ElT}}
+  #   return new{ElT,DataT}(data)
+  # end
 
   # Special case of Array which gets turned into a Vector
-  function Dense{ElT,VecT}(data::Array) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(vec(data))
-  end
+  # function Dense{ElT,DataT}(data::Array) where {ElT,DataT<:AbstractArray{ElT}}
+  #   return new{ElT,DataT}(vec(data))
+  # end
+end
+
+#Start with high information constructors and move to low information constructors
+function Dense{ElT,DataT}() where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(DataT())
+end
+
+# Construct from a set of indices
+# This will fail if zero(ElT) is not defined for the ElT
+function Dense{ElT,DataT}(inds::Tuple) where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(zeros(DataT, dim(inds)))
```

Comment:
I guess this should use `generic_zeros`? Would the plan be to use `generic_zeros(Dense{ElT,DataT}, dim(inds))` for this kind of thing?

Reply:
I think either calling `generic_zeros` with datatype or storagetype here would be fine. My thought was that someone somewhere else might just want to do something like
```
Function f(eltype::Type{<:Number>}, datatype::Type{<:AbstractArray{eltype}})
DenseDataStruct = Dense{eltype, datatype}
d = generic_zeros(DenseDataStruct)
### Do work with d
end
```